compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -421,6 +517,33 @@ pub(super) fn try_unify_abstract_consts<'tcx>(
     // on `ErrorReported`.
 }
 
+fn walk_abstract_const<'tcx, F>(tcx: TyCtxt<'tcx>, ct: AbstractConst<'tcx>, mut f: F)
```

Comment:
is there any particular reason we need to walk this in tree order and not just visit all nodes?

Reply:
So for subtrees that would be incorrect, as they can contain nodes from other subtrees.

I think if we always use a method like this for a complete tree that would be fine. Kinda afraid that we accidentially incluse unused nodes though :thinking: 

About unused nodes, do we want to emit an error if we encounter them during `AbstractConst` building.
Let me test how this deals with things like `N + 1; 3` rn.

Reply:
```rust
#![feature(const_generics, const_evaluatable_checked)]

fn test<const N: usize>() -> usize {
    N; 2
}
```
results in
```
fn test() -> usize {
    let mut _0: usize;                   // return place in scope 0 at ./example.rs:3:30: 3:35
    let _1: usize;                       // in scope 0 at ./example.rs:4:5: 4:6

    bb0: {
        StorageLive(_1);                 // scope 0 at ./example.rs:4:5: 4:6
        _1 = const N;                    // scope 0 at ./example.rs:4:5: 4:6
        StorageDead(_1);                 // scope 0 at ./example.rs:4:6: 4:7
        _0 = const 2_usize;              // scope 0 at ./example.rs:4:8: 4:9
        return;                          // scope 0 at ./example.rs:5:2: 5:2
    }
}
```
which is currently allowed. Do we want to forbid unused nodes in abstract const building?